text/0023-stats-component.md

```diff
@@ -0,0 +1,425 @@
+  error - the number of cells in that column that contain invalid boolean values
+  missing - the number of cells in that column that do not contain data 
+  trueCount - the number of cells that contain a true value
+  falseCount - the number of cells that contain a false value
+
+The structure of the default stats component depends on the structure of the data itself. If each row of the dataset body is expressed as an array, the stats component will be an array of objects. If each row of the dataset body is expressed as an object, the stats component will be an object of objects. One important thing to note, we must have a structure with a schema that describes the title and the type in order to create stats.
+
+Let's look at two examples of a small dataset body, and the resulting stats object.
+
+```json
+// this is a small dataset body expressed as a 2D array. It is a list of 
+// classes in an elementary school. The columns are the class name (string),
+// the current number of students (number), and whether the class can accept 
+// more students
+[
+  ["1A", 20, false],
+  ["1B", 22, false],
+  ["1C", 17, true],
+  ["1D", 19, true]
+]
+
+// the resulting stats section would look like this:
+[
+  {
+    "title":"class name",
+    "type":"string",
+    "count": 4,
+    "minLength": 2,
+    "maxLength": 2
+  },
+  {
+    "title":"students",
+    "type":"number",
+    "count": 4,
+    "min": 17,
+    "max": 22,
+    "avg": 19.5,    
+  },
+  {
+    "title":"can accept more students",
+    "type":"boolean",
+    "count": 4,
+    "trueCount": 2,
+    "falseCount": 2,
+  }
+]
+```
+Note that the data we expressed as a 2D array, and so the stats component is presented as an array.
+
+Let's look at this same example, but with the data presented as an array of objects:
+
+```json
+[
+  {
+    "class_name":"1A",
+    "students":20, 
+    "can_accept_more":false
+  },
+  {
+    "class_name":"1B",
+    "students":22, 
+    "can_accept_more":false
+  },
+  {
+    "class_name":"1C",
+    "students":27, 
+    "can_accept_more":true
+  },
+  {
+    "class_name":"1D",
+    "students":19, 
+    "can_accept_more_students":true
+  }
+]
+
+// stats:
+{
+  "class_name": {
+    "title":"class name",
+    "type":"string",
+    "count": 4,
+    "minLength": 2,
+    "maxLength": 2
+  },
+  "students": {
+    "title":"students",
+    "type":"number",
+    "count": 4,
+    "min": 17,
+    "max": 22,
+    "avg": 19.5,    
+  },
+  "can_accept_more_students": {
+    "title":"can accept more students",
+    "type":"boolean",
+    "count": 4,
+    "trueCount": 2,
+    "falseCount": 2,
+  }
+}
+```
+
+In summary, qri can generate some basic stats on a dataset body, if the body is tabular. We also need a structure with a schema in order to understand what stats to be calculating.
+
+## high dimentional data
+Not only does 3 dimentinal (or higher) data have a less apparent structure, it is also diffcult to reason about the best way to display stats for 3 dimentional data.
```

Comment:
This section on high dimensional data feels less worked out. I think we should drop it for now, save the writing, and work on it in a follow-up RFC. Instead I think "high dimensional data" should just focus on stubbing out _how high dimensional statistics can be expressed_

> Calculating statistics on high dimensional data is out of scope for this RFC. however we do provide a way to arrange statistical metadata into hierarchies by nesting stats object with a key: `stats`. Given a sample of high-dimensional data: 

```
[
    {
      "type": "Feature",
      "properties": {
        "name":"Brooklyn"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [
          -74.00,
          40.63
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "name":"Manhattan"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [
          -74.00,
          40.71
        ]
      }
    }
  ]
```

Here's an example of "stats that contain stats"

```js
{
  // other dataset components ...
  "stats":  [
    { "title": "type", "type" : "string" /* same as 2D string stat */ },
    {
       "title": "geometry",
       "type" : "object",
        "error" : 0,
        "valid" : 2,
        "missing" : 0, 
        "stats": {
          "geometry" : { "type" : "string" /* same as 2D string stat */ } ,
          "coordinates": {
             "type" : "geopoint",
             "avg" : [-74.00, 41.00]
          }
        }
    }
  }
}
```

What's important to note here is stats is a _recursive data structure_ that uses the keyword "stats" to build up statistical hierarchies. The actual calculation of these statistics is out of scope, but their presentation can be validated, mainly by reserving the `stats` keyword within a stats object.